### PR TITLE
Link gamecore binaries to correct runtime libs

### PR DESCRIPTION
**Pull request by thhous-msft**
On console, gamecore is required to be linked to the console specific umbrella library. This library is not included with the public GDK however, so we will never be able to build gamecore specific binaries in public CI.

On the Windows 11 SDK, WINAPI_FAMILY_GAMES does not work anymore to compile, even with the ReadNoFence hack. Other functions, including some schannel functions are gone as well. So converted to the same trick on UWP, where we say we're a desktop app, but we only link to the proper umbrella libraries.  This let me join the preprocessor definitions and create a restricted build category for both UWP and Gamecore.

Investigating looking into onebranch to build gamecore binaries in internal CI.

### CODE REVIEW

File: src/inc/quic_platform_winuser.h [hunk 1]
```diff
@@ -28,7 +28,7 @@
 #define WIN32_LEAN_AND_MEAN 1
 #endif
 
-#ifdef QUIC_UWP_BUILD
+#if defined(QUIC_RESTRICTED_BUILD)
 #undef WINAPI_FAMILY
 #define WINAPI_FAMILY WINAPI_FAMILY_DESKTOP_APP
```
[the remark this answers is not in the document]
Comment (thhous-msft) [reply src/inc/quic_platform_winuser.h:33]: Its not just ReadNoFence thats blocked. Schannel API's are blocked as well. Cmake actually always passes WINAPI_FAMILY_DESKTOP_APP, which is why previously it built.

File: .azure/azure-pipelines.ci.yml [hunk 2]
```diff
@@ -209,13 +209,14 @@ stages:
...
       tls: openssl
       extraBuildArgs: -DisableTools -DisableTest -EnableTelemetryAsserts
-  - template: ./templates/build-config-user.yml
-    parameters:
-      image: windows-latest
-      platform: gamecore
-      arch: x64
-      tls: schannel
-      extraBuildArgs: -DisableTools -DisableTest -EnableTelemetryAsserts
+  # Gamecore will not be able to be built in public CI
```
Comment (nibanks) [inline .azure/azure-pipelines.ci.yml:212]: Is the because the GDK isn't available? If so, then https://github.com/actions/virtual-environments/issues/4053 should unblock that (eventually).

File: CMakeLists.txt [hunk 3]
```diff
@@ -364,11 +364,41 @@ if(WIN32)
...
 
     if (QUIC_GAMECORE_BUILD)
-        list(APPEND QUIC_COMMON_DEFINES WINAPI_FAMILY=WINAPI_FAMILY_GAMES QUIC_GAMECORE_BUILD)
+        list(APPEND QUIC_COMMON_DEFINES WINAPI_FAMILY=WINAPI_FAMILY_DESKTOP_APP QUIC_RESTRICTED_BUILD)
+        set(CMAKE_CXX_STANDARD_LIBRARIES "")
+        set(CMAKE_CXX_STANDARD_LIBRARIES_INIT "")
+        set(CMAKE_C_STANDARD_LIBRARIES "")
+        set(CMAKE_C_STANDARD_LIBRARIES_INIT "")
+
+        set(UnsupportedLibs advapi32.lib comctl32.lib comsupp.lib dbghelp.lib gdi32.lib gdiplus.lib guardcfw.lib kernel32.lib mmc.lib msimg32.lib msvcole.lib msvcoled.lib mswsock.lib ntstrsafe.lib ole2.lib ole2autd.lib ole2auto.lib ole2d.lib ole2ui.lib ole2uid.lib ole32.lib oleacc.lib oleaut32.lib oledlg.lib oledlgd.lib oldnames.lib runtimeobject.lib shell32.lib shlwapi.lib strsafe.lib urlmon.lib user32.lib userenv.lib wlmole.lib wlmoled.lib onecore.lib)
```
Comment (nibanks) [inline CMakeLists.txt:381]: Where did you get this list? Why is `onecore.lib` on it?

File: CMakeLists.txt [hunk 4]
```diff
@@ -394,8 +424,8 @@ if(WIN32)
     # These cannot be updated until CMake 3.13
     set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} /GL /Zi")
     set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} /GL /Zi")
-    set(CMAKE_SHARED_LINKER_FLAGS_RELEASE "${CMAKE_SHARED_LINKER_FLAGS_RELEASE} /LTCG /DEBUG /OPT:REF /OPT:ICF")
-    set(CMAKE_EXE_LINKER_FLAGS_RELEASE "${CMAKE_EXE_LINKER_FLAGS_RELEASE} /LTCG /DEBUG /OPT:REF /OPT:ICF")
+    set(CMAKE_SHARED_LINKER_FLAGS_RELEASE "${CMAKE_SHARED_LINKER_FLAGS_RELEASE} /LTCG /IGNORE:4075 ${Console_ArchOptions_LTCG} /DEBUG /OPT:REF /OPT:ICF")
+    set(CMAKE_EXE_LINKER_FLAGS_RELEASE "${CMAKE_EXE_LINKER_FLAGS_RELEASE} /LTCG /IGNORE:4075 ${Console_ArchOptions_LTCG} /DEBUG /OPT:REF /OPT:ICF")
```
Comment (nibanks) [inline CMakeLists.txt:428]: What does this do?

File: .azure/azure-pipelines.ci.yml (hunk 2 above)
Comment (thhous-msft) [reply .azure/azure-pipelines.ci.yml:212]: It will never work because it requires private gamecore.

File: CMakeLists.txt (hunk 3 above)
Comment (thhous-msft) [reply CMakeLists.txt:381]: https://github.com/microsoft/Xbox-GDK-Samples/blob/main/Samples/Tools/CMakeExample/XboxOne/CMakeGDKXboxOne.cmake#L199

File: CMakeLists.txt (hunk 4 above)
Comment (thhous-msft) [reply CMakeLists.txt:428]: Ignores the incremental warning, and then ignores a specific flag for the scarlett build.

File: .azure/azure-pipelines.ci.yml (hunk 2 above)
Comment (nibanks) [reply .azure/azure-pipelines.ci.yml:212]: If this doesn't work, just delete it instead of commenting it out.

File: src/bin/CMakeLists.txt [hunk 5]
```diff
@@ -234,7 +234,12 @@ set_property(TARGET msquic PROPERTY FOLDER "${QUIC_FOLDER_PREFIX}libraries")
 
 if(WIN32)
     if(QUIC_UWP_BUILD)
-        target_link_libraries(msquic PUBLIC OneCoreUAP)
+        target_link_libraries(msquic PUBLIC OneCore)
```
Comment (nibanks) [inline src/bin/CMakeLists.txt:237]: Why this change?

Comment (ThadHouse) [reply src/bin/CMakeLists.txt:237]: I need to revert this actually. I'll do it next time I'm at a computer.

File: CMakeLists.txt [hunk 6]
```diff
@@ -364,11 +364,41 @@ if(WIN32)
...
+        if (QUIC_SCARLETT_BUILD)
+            list(APPEND Console_ArchOptions /arch:AVX2 /d2vzeroupper)
+            set(Console_ArchOptions_LTCG /d2:-vzeroupper)
+        else()
+            list(APPEND Console_ArchOptions /arch:AVX)
+        endif()
+
+        # Locate Software Development Kits
+        GET_FILENAME_COMPONENT(Console_SdkRoot "[HKEY_LOCAL_MACHINE\\SOFTWARE\\Microsoft\\GDK;InstallPath]" ABSOLUTE CACHE)
+        set(XdkEditionTarget "211000")
```
[the remark this answers is not in the document]
Comment (thhous-msft) [reply CMakeLists.txt]: Done

File: CMakeLists.txt [hunk 7]
```diff
@@ -364,11 +364,41 @@ if(WIN32)
...
+
+        if (QUIC_SCARLETT_BUILD)
+            list(APPEND Console_ArchOptions /arch:AVX2 /d2vzeroupper)
+            set(Console_ArchOptions_LTCG /d2:-vzeroupper)
+        else()
+            list(APPEND Console_ArchOptions /arch:AVX)
+        endif()
+
+        # Locate Software Development Kits
+        GET_FILENAME_COMPONENT(Console_SdkRoot "[HKEY_LOCAL_MACHINE\\SOFTWARE\\Microsoft\\GDK;InstallPath]" ABSOLUTE CACHE)
```
[the remark this answers is not in the document]
Comment (thhous-msft) [reply CMakeLists.txt]: Done

**Review by nibanks — APPROVED** [review]

**Review by DougGal-Microsoft — COMMENTED** [review]
Looks good to me, thank you!
